packages/keyring/src/keyring.ts

```diff
@@ -0,0 +1,319 @@
+
+    console.log('Mnemonic leaked to memory as a string while importing wallet. This is expected.')
+
+    const mnemonic = mnemonicStringToUint8Array(mnemonicStr)
+
+    this.clearCachedSecrets()
+    this._initFromMnemonic(mnemonic, '')
+
+    return mnemonic
+  }
+
+  public initFromEncryptedMnemonic = (encryptedMnemonic: string, password: string, passphrase: string) => {
+    const { version, decryptedMnemonic } = decryptMnemonic(encryptedMnemonic, password)
+
+    this.clearCachedSecrets()
+    this._initFromMnemonic(decryptedMnemonic, passphrase)
+
+    encryptedMnemonic = ''
+    password = ''
+    passphrase = ''
+    resetArray(decryptedMnemonic)
+
+    return version
+  }
+
+  public generateAndCacheAddress = (props: GenerateAddressProps): NonSensitiveAddressData =>
+    this._getNonSensitiveAddressData(this._generateAndCacheAddress(props))
+
+  public signTransaction = (txId: string, addressHash: AddressHash): string =>
+    transactionSign(txId, this.exportPrivateKeyOfAddress(addressHash))
+
+  public signMessage = (message: string, addressHash: AddressHash): string =>
+    sign(message, this.exportPrivateKeyOfAddress(addressHash))
+
+  public exportPrivateKeyOfAddress = (addressHash: AddressHash): string =>
+    bytesToHex(this._getAddress(addressHash).privateKey)
+
+  public discoverAndCacheActiveAddresses = async (
+    client: ExplorerProvider,
+    addressIndexesToSkip: number[] = [],
+    minGap = 5
+  ): Promise<NonSensitiveAddressData[]> => {
+    const addressesPerGroup = Array.from({ length: TOTAL_NUMBER_OF_GROUPS }, (): SensitiveAddressData[] => [])
+    const activeAddresses: SensitiveAddressData[] = []
+    const skipIndexes = Array.from(addressIndexesToSkip)
+
+    for (let group = 0; group < TOTAL_NUMBER_OF_GROUPS; group++) {
+      const newAddresses = this._deriveAddressesInGroup(group, minGap, skipIndexes)
+      addressesPerGroup[group] = newAddresses
+      skipIndexes.push(...newAddresses.map((address) => address.index))
+    }
+
+    const addressesToCheckIfActive = addressesPerGroup.flat().map((address) => address.hash)
+    const results = await getActiveAddressesResults(addressesToCheckIfActive, client)
+    const resultsPerGroup = splitResultsArrayIntoOneArrayPerGroup(results, minGap)
+
+    for (let group = 0; group < TOTAL_NUMBER_OF_GROUPS; group++) {
+      const { gap, activeAddresses: newActiveAddresses } = getGapFromLastActiveAddress(
+        addressesPerGroup[group],
+        resultsPerGroup[group]
+      )
+
+      let gapPerGroup = gap
+      activeAddresses.push(...newActiveAddresses)
+
+      while (gapPerGroup < minGap) {
+        const remainingGap = minGap - gapPerGroup
+        const newAddresses = this._deriveAddressesInGroup(group, remainingGap, skipIndexes)
+        skipIndexes.push(...newAddresses.map((address) => address.index))
+
+        const addressesToCheckIfActive = newAddresses.map((address) => address.hash)
```

Comment:
NIT: I'd rename this to `newAddressesToCheckIfActive` to make sure it's not confused with the variable with the same name above.

Reply:
Addressed in d39c6f560 in PR https://github.com/alephium/alephium-frontend/pull/504